### DISCUSSION TITLE
v4.0.41 pre-release

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#105] [General]
Enhancements, and fixes.

* Fix: The integration was still messing up where the config folder is sometimes. Hopefully really fixed.
* Fix: I sheepishly admit that I broke forecast power now/+30 mins/+60 mins in v4.0.40.
* Enhancement: Consistency in state_class for the power now/+30/+60 sensors.
* Enhancement: More half hour interpolation goodness, where splines for the entire day are calculated at each forecast get (and at midnight) and are referenced without calculation later. This work also solves for potential 'spline bounce', where a forecast remaining could exceed forecast for the day, or values dipped below zero at the end of the day (or bizarrly went -0.0, which the floating-point IEEE committee can explain what they were thinking on that front...) 

As always, both positive and negative feedback is most welcome, and this will contribute to getting the release ultimately published as latest. Report back, and don't be shy!

Feedback in the past has been gratefully received from @dtswk, @EddieSpaghettie, @ffm777, @gcoan, @hwb45, @iainfogg, @isorin, @jeffb1974, @kernow, @ladaan, @Markyt77, @nilogax, @Noodleyman, @ProphetOfDoom, @rholligan, @Unlik and @wimdebruyn, and continued thanks to them for their interest in beta releases (and alphas!) and for their contributions.

> [!NOTE]
> The BJReplay fork is **_still_** not published for searching in HACS, so continued use of a manual install in HACS is the go. As explained earlier we're going to keep any releases as pre-release on the fork until that happens, as any 'latest' version change would require modification of that HACS pull request.

## What's Changed
* Interpolated forecast now/+30/+60 fix #101 by @autoSteve
* Ensure config directory is always relative to install location #98 by @autoSteve
* Add state_class to `power_now_30m` and `power_now_1hr` to match `power_now` by @autoSteve (will remove LTS, but LTS is not useful for these sensors)
* Utilise daily splines of momentary and reducing forecast values by @autoSteve

Full Changelog: https://github.com/BJReplay/ha-solcast-solar/compare/v4.0.40...v4.0.41

**Comment by autoSteve** [comment]
So    👏    👏    👏...    😑 Meh?    Or 🤯🤷‍♂️ how does it do that? Or 🤬?

I'll just chat amongst myself.

I know that twelve wonderful humans other than myself are running this code. Hit me up once forecasts have happened, midnight has occurred, splines have presented themselves in all their glory, and hopefully nothing has gone 'pop'.

**Comment by Nilogax** [comment]
Progress so far : Installed OK, not apparently broken anything yet, no errors in log, haven't had a chance to look at any spliney goodness.   Oh and forecasts have been updated. Let's see if it turns into a pumpkin at midnight...

**Reply by autoSteve** [reply]
I take it that at midnight your integration didn't turn back into a pumpkin @Nilogax, nor your horses to mice, your footmen to lizards, and your clothes back into rags again? I'm hoping that a big pile of wonderful nothing happened, and that the fairy godmother was a worry wort, and you're still kicking on after the ball at a club somewhere...

**Reply by Nilogax** [reply]
Happy to report that nothing even the slightest bit cataclysimic has happened.

**Comment by gcoan** [comment]
Sorry for the no updates from me, I am away for 2 weeks helping to marshal the World Microlight Championships (www.wmc2024.com).

About half way through the first week I lost remote connectivity to my Home Assistant server, I think its not running but don't know why so won't be able to test Solcast until I get back.

Fortunately the inverters are still running OK and in Eco mode so generating lots and running the house off free solar every night

**Reply by gcoan** [reply]
Just installed 0.41 today (jumping from 0.39 and missing out 0.40).

Installed OK, no errors in the log, the sensors appear all populated OK.

I did see a slight bounce in the forecast_pv_remaining_today, but that was due to the solcast update automation running at 13:00 and getting a slightly better pv_today forecast from solcast.

If you can work out a way to predict and prevent this bounce I'll be very impressed.....
(and of course the bounce could be up or down, just to make the challenge harder)

![image](https://github.com/user-attachments/assets/15e54d4d-9460-4159-9a9c-944cb50e1928)

**Reply by autoSteve** [reply]
Yeah, nah. Forecast get is outside the bounce remit. That will happen due to Solcast shifting the goals, and is impossible to predict and smooth, given it's a total value re-write, not a weird statistics bounce thing. (Your other spliney goodness looks groovy, but.)

Good job skipping v4.0.40. Much self-shame with one aspect of that release. Good riddance.

**Comment by ProphetOfDoom** [comment]
Just jumped onto V4.0.41. All sensors populated OK & no errors in logs, but won't get any real values until my automation runs at sunset tomorrow morning.
In the meantime, two observations/questions:
O1. Spotted a "rendomisation" typo in the readme
Q1. Hard Limit Set. The documentation is a bit woolly about what this is. It reads like it's an Inverter ac limit, so if your PV wants to push out 6kW dc, but you have a 5kW ac inverter, or even a 6kW ac inverter that has been dialled down to 5kW ac at the direction of the local DNO (reduction in registered capacity - not the same as an export limit), then you are going to get clipping occur. Does the code for this actually work, and does anyone else concur that it could be better named/described

[ For non-UK readers the DNO is the Distibution Network Operator, the people who install and maintain the electricity supply network in your local area. This is not the same as your Electricity Supplier (the people you actually buy your electricity from, that travels over the network owned and operated by the DNO). Inverter capacity above 3.68kW has to be pre-approved by the DNO, and for larger installations in areas with poor infrastructure, there is a chance that the DNO will not only impose an export limit, but also limit your total installed inverter capacity in order to protect the local network. ]

**Reply by autoSteve** [reply]
The random "rendom" is fixed on my clone. Thanks for that @ProphetOfDoom.

I do agree that the hard limit doco is sparse. Oziee's release changes for v4.0.22 probably explain why.

> added a config for setting a hard limit for inverters with over sized solar arrays *99.9999999% of users will not need to ever user and set this (0.00000001% is @CarrapiettM)

It's exact function in the real world bends my mind a bit. Maybe @CarrapiettM can shed some light.

When one's Solcast account is configured both DC and AC capacity values are specified, and this makes sense to me. The forecasted values for a site should fall under the AC limit (inverter output). Why these would need to be further clipped in the integration is the bit that bends my head.

**Reply by ProphetOfDoom** [reply]
@autoSteve The ac capacity limits in Solcast are per roof. If you have a 5kW inverter and 3kW of PV on each of two roofs, you might be tempted to set the ac capacity limit for each roof to 2.5kW. But if you do this, Solcast might prematurely limit one roof to 2.5kW when the other is only producing 1kW. Conversely, if you set the ac limit for each roof to the PV dc capacity of 3kW (no point in setting it higher), you might end up with 6kW into a 5kW inverter, which would likely cause clipping in the real world (need to check exactly how a hybrid behaves in this situation)

I think this is only a thing for PV arrays connected to String/Hybrid inverters with/without optimisers, as micro-inverters have a per-panel hard limit, so the roof limit is easily quantifiable.

**Reply by autoSteve** [reply]
Makes sense. Micro-inverters here, so I've not played with strings.

**Reply by autoSteve** [reply]
This begs a big question, @ProphetOfDoom.

Should the hard limit be more easily configurable than it is at the moment? i.e would it apply to more users than the 0.00000001% that Oziee stated? I would not think so, because typical PV installs.

I'm having a hard time getting my head around how one would configure two Solcast rooftops that face different directions yet share an inverter. A hard time mostly because I don't know the ins-and-outs of how Solcast utilise the DC/AC capacity settings. That and I have micro-inverters so have never had to think about it.

I know if the AC capacity set is too low then forecast 'clipping' can occur, as stated on their site. I have also seen the word "ratio" mentioned in the doco, but have not seen anything that explains how that ratio applies. Are power figures calculated using a ratio of these two capacity values? For a system with two strings facing different directions, which utilise a single inverter it could be configured as a single rooftop with azimuth/tilt averaged. (And if so then the hard limit setting would not be used.) Or if configured as two rooftop sites then what AC value would be chosen for each? (I would imagine a proportionate share of the inverter maximum, where the hard limit setting would not be used.)

But then my brain breaks when considering over-sized DC string systems, which is where I think the hard limit might enter the picture. Would one configure a Solcast AC value higher than actual (so the forecast is _not_ clipped), and then 'clip' to a maximum using the integration hard limit? (But then what of the ratio of DC to AC hinted at Solcast?)

**Reply by autoSteve** [reply]
And to answer one of your original questions, "Does it work?" The answer is yes, and no.

For the sum of all rooftop sites it will work.

```
itm["pv_estimate"] = min(round(itm["pv_estimate"] + (x["pv_estimate"] * self._damp[h]),4), self._hardlimit)
```

However, it won't work for the sites breakdown attributes that @gcoan utilises. The same min() function is used, so each site breakdown vs. the hard limit set will be non-sensical, as the hard limit will be much higher than the individual site maximum. (Sorry if you set a hard limit, @gcoan, but if so it's something to be aware of.) I'm not sure how this could be improved, short of separate site hard limits. I also don't know whether it even needs to be improved, because 0.00000001%. I guess a ratio of the hard limit in line with each rooftop expected production could be used. Maybe.

**Comment by Unlik** [comment]
For me **all good with v4.0.41**, no new issues in logs, and olds are gone. Many thanks.

**Comment by autoSteve** [comment]
A perfectly calculated reducing spline for me today. No -0.0 crap at the end of the day, forecast remaining always at <= total forecast for the day. I'm happy with that.

I'm also happy that there are currently 53 other people quietly enjoying this beta code.

I have observed one oddity around midnight local time, and I think I can explain this. The remaining for today took a jump at 00:05. This doesn't bother anyone but me, but I think the per-5-minute sensor values managed to recalculate before the daily spline recalculation occurred (logs are gone because restarts, so I cannot confirm yet). I am investigating whether this can be tightened up, but I'm not absolutely sure it needs tightening up, because midnight. (The other minor bump some hours later is a forecast update in the wee hours.)

<img width="379" alt="image" src="https://github.com/user-attachments/assets/84e9cb29-4279-498f-b81f-ea6ea854ba1d">

**Reply by autoSteve** [reply]
A patch for the midnight oddity is in v4.0.next. This will require further testing to make sure it's fixed.

91 downloads with barely a squeak of any issue.

**Comment by autoSteve** [comment]
There is also some minor spline bounce with the forecast 0/+30/+60 sensors, which is not unsurprising given spline bounce is not considered at the moment for the "moment" splines. It's a mere annoyance, but wants looking into at some stage.

<img width="555" alt="image" src="https://github.com/user-attachments/assets/d9cbbf1b-fc36-443c-a137-106344c1417f">

**Reply by autoSteve** [reply]
Moment spline bounce is already suppressed in v4.0.next.

**Comment by jeffb1974** [comment]
Likewise.  Only really using the daily solar total (for two split arrays)
but working well.

**Comment by autoSteve** [comment]
Known issues and fixes completed so far, 148 downloads to date.

* Doco updates. Hard limit described #109, typos, more explanations, tips.
* 'Moment' splines have had 'spline bounce' suppressed.
* A midnight occasional race to update splines before sensors has been fixed. An annoyance, but it made my eye twitch.
* New installations that fail the initial sites 'get' with a 429 will continuously retry forever until sites are acquired (new users will not thank me, but indirectly actually will thank me by not raising an issue).
* Code tidying has happened, not impacting function just fixing my OCD.
* For now I have given up trying to apply dampening and hard limit to site breakdown attributes. These are now completely removed for sites breakdown attributes. The only way to solve this for these attributes is to make dampening and hard limit site specific. Which would be cool, but way too complex. (Or code a site-specific sensor to do it yourself.) It is what it is for now.

**Comment by dtswk** [comment]
Please excuse me, currently on holiday in Europe. Back in 3 weeks. :)

**Comment by autoSteve** [comment]
And so slips v4.0.41 into the night. Long live v4.0.42.

If you grok the world of Douglas Adams' Hitchhikers, 42 is _the answer_, so that's a good omen.
